File: src/app/server/response_type.py
import logging

from py_expression_eval import Parser
from voluptuous import All, Any, Coerce, Length, Optional, Required, Schema


log = logging.getLogger('app.response_type')

response_parts_schema = Schema([
    {
        # Common fields
        Required('id', default=None): Any(
            All(str, Length(min=1)), None),
        'type': Any('multiple_choice', 'numerical'),
        Required('name', default=None): Any(
            All(str, Length(min=1)), None),
        Required('description', default=None): Any(
            All(str, Length(min=1)), None),
        # Fields that vary by type
        # Multiple choice
        Optional('options'): All([
            {
                'score': Coerce(float),
                'name': All(str, Length(min=1)),
                Required('if', default=None): Any(
                    All(str, Length(min=1)), None),
                Required('description', default=None): Any(
                    All(str, Length(min=1)), None)
            }
        ], Length(min=2)),
        # Numerical
        Optional('lower'): Any(All(str, Length(min=1)), None),
        Optional('upper'): Any(All(str, Length(min=1)), None),
    },
], required=True)


response_schema = Schema([
    Any(
        {
            'index': int,
            'note': All(str, Length(min=1)),
        },
        {
            'value': Coerce(float),
        },
    )
], required=True)


class ResponseTypeError(Exception):
    pass


class ExpressionError(ResponseTypeError):
    '''Error in response type definition'''
    pass


class ResponseError(ResponseTypeError):
    '''Error in user-provided response'''
    pass


# These functions are used to calculate scores and validate entered data.
# NOTE: keep changes made to these classes in sync with those in response.js.


parser = Parser()


def unique_strings(ss):
    return sorted(set(ss))


def parse(exp):
    try:
        return parser.parse(exp) if exp else None
    except Exception as e:
        raise ExpressionError("'%s': %s" % (exp, e))


def refs(c_exp):
    return c_exp.variables() if c_exp else []


class ResponseType:
    def __init__(self, name, parts_def, formula):
        self.name = name
        self.parts = [response_part(p_def) for p_def in parts_def]
        self.formula = parse(formula)
        self.declared_vars = unique_strings([
            v for part in self.parts
            for v in part.declared_vars])
        self.free_vars = unique_strings([
            v for part in self.parts
            for v in part.free_vars] + refs(self.formula))
        self.unbound_vars = unique_strings([
            v for v in self.free_vars
            if v not in self.declared_vars])

    def score(self, response_parts, scope):
        if self.formula is not None:
            try:
                return self.formula.evaluate(scope)
            except Exception as e:
                raise ExpressionError(str(e))

        score = 0.0
        for part_t, part_r in zip(self.parts, response_parts):
            try:
                score += part_t.score(part_r)
            except Exception as e:
                i = self.parts.index(part_t)
                raise ResponseError(
                    "Response part %d is invalid: %s" % (i, e))
        return score

    def variables(self, response_parts, ignore_errors=False):
        scope = {}
        for part_t, part_r in zip(self.parts, response_parts):
            try:
                scope.update(part_t.variables(part_r))
            except Exception as e:
                if ignore_errors:
                    i = self.parts.index(part_t)
                    raise ResponseError(
                        "Response part %d is invalid: %s" % (i, e))
        return scope

    def validate(self, response_parts, scope):
        for i, (part_t, part_r) in enumerate(zip(self.parts, response_parts)):
            try:
                part_t.validate(part_r, scope)
            except Exception as e:
                if part_t.name:
                    name = part_t.name
                    if len(self.parts) > 1:
                        name += " (part %d)" % (i + 1)
                else:
                    if len(self.parts) > 1:
                        name = "Response part %d" % (i + 1)
                    else:
                        name = "Response"
                raise ResponseError(
                    "%s is invalid: %s" % (name, e))

    def calculate_score(self, response_parts):
        if response_parts is None or len(response_parts) != len(self.parts):
            raise ResponseError("Response is incomplete")

        scope = self.variables(response_parts)
        self.validate(response_parts, scope)
        return self.score(response_parts, scope)

    def __repr__(self):
        return "ResponseType(%s)" % (self.name)


def response_part(p_def):
    if p_def['type'] == 'multiple_choice':
        return MultipleChoice(p_def)
    else:
        return Numerical(p_def)


class ResponsePart:
    def __init__(self, p_def):
        self.id_ = p_def.get('id')
        self.name = p_def.get('name')
        self.type = p_def['type']
        self.description = p_def.get('description')


class MultipleChoice(ResponsePart):
    def __init__(self, p_def):
        super().__init__(p_def)
        self.options = [ResponseOption(o_def)
                        for o_def in p_def['options']]
        self.declared_vars = [self.id_, self.id_ + '__i'] if self.id_ else []
        self.free_vars = unique_strings([
            v for opt in self.options
            for v in opt.free_vars])

    def get_option(self, part_r):
        try:
            i = part_r['index']
        except KeyError:
            raise ResponseError("No index specified")
        if i < 0:
            raise IndexError("Option %d does not exist" % i)
        option = self.options[i]
        return i, option

    def score(self, part_r):
        _, option = self.get_option(part_r)
        return option.score

    def variables(self, part_r):
        variables = {}
        if self.id_:
            i, option = self.get_option(part_r)
            variables[self.id_] = option.score
            variables[self.id_ + '__i'] = i
        return variables

    def validate(self, part_r, scope):
        _, option = self.get_option(part_r)
        if not option.available(scope):
            raise ResponseError(
                "Conditions for option '%s' are not met" % option.name)

    def __repr__(self):
        return "MultipleChoice(%s)" % (self.name or self.id_)


class ResponseOption:
    def __init__(self, o_def):
        self.score = o_def['score']
        self.name = o_def['name']
        self.predicate = parse(o_def.get('if'))
        self.free_vars = refs(self.predicate)

    def available(self, scope):
        if not self.predicate:
            return True

        try:
            return bool(self.predicate.evaluate(scope))
        except Exception:
            return False

    def __repr__(self):
        return "ResponseOption(%s: %f)" % (self.name, self.score)


class Numerical(ResponsePart):
    def __init__(self, p_def):
        super().__init__(p_def)
        self.lower_exp = parse(p_def.get('lower'))
        self.upper_exp = parse(p_def.get('upper'))
        self.declared_vars = [self.id_] if self.id_ else []
        self.free_vars = unique_strings(
            refs(self.lower_exp) + refs(self.upper_exp))

    def score(self, part_r):
        try:
            return part_r['value']
        except KeyError:
            raise ResponseError("No value specified")

    def variables(self, part_r):
        variables = {}
        if self.id_:
            variables[self.id_] = part_r['value']
        return variables;

    def lower(self, scope):
        if not self.lower_exp:
            return float('-inf')
        try:
            return self.lower_exp.evaluate(scope)
        except Exception as e:
            log.debug('Could not evaluate lower_exp: %s: %s',
                      self.lower_exp.toString(), e)
            raise ExpressionError(str(e))

    def upper(self, scope):
        if not self.upper_exp:
            return float('inf')
        try:
            return self.upper_exp.evaluate(scope)
        except Exception as e:
            log.debug('Could not evaluate upper_exp: %s: %s',
                      self.upper_exp.toString(), e)
            raise ExpressionError(str(e))

    def validate(self, part_r, scope):
        score = self.score(part_r)
        if self.lower(scope) > score:
            raise ResponseError("Value must be greater than %s" %
                                self.lower(scope))
        if self.upper(scope) < score:
            raise ResponseError("Value must be less than %s" %
                                self.upper(scope))

    def __repr__(self):
        return "Numerical(%s)" % (self.name or self.id_)

File: src/app/client/js/response-type.js
'use strict';

angular.module('upmark.response.type', [
    'ngResource', 'upmark.admin.settings', 'upmark.user', 'upmark.chain'])


.config(function($routeProvider, chainProvider) {
    $routeProvider
        .when('/:uv/response-type/:responseType', {
            templateUrl : 'response_type.html',
            controller : 'ResponseTypeCtrl',
            resolve: {routeData: chainProvider({
                responseType: ['ResponseType', '$route',
                        function(ResponseType, $route) {
                    return ResponseType.get({
                        id: $route.current.params.responseType,
                        programId: $route.current.params.program
                    }).$promise;
                }]
            })}
        })
    ;
})


.factory('responseTypes', function() {

    // These functions are used to calculate scores and validate entered data.
    // NOTE: keep changes made to these classes in sync with those in
    // response_type.py.

    var uniqueStrings = function(ss) {
        var set = {},
            res = [];
        ss.forEach(function(s) {
            if (set.hasOwnProperty(s))
                return;
            set[s] = true;
            res.push(s);
        });
        return res.sort();
    };
    var parse = function(exp) {
        try {
            return exp ? Parser.parse(exp) : null;
        } catch (e) {
            if (/parse error/.exec(e))
                console.log("Failed to parse expression '" + exp + "': " + e)
            return null;
        }
    };
    var refs = function(cExp) {
        return cExp ? cExp.variables() : [];
    };




    /*function ResponseType(name, partsDef, formula, merge) {
        this.name = name;
        this.parts = partsDef && partsDef.map(responsePart) || [];s
        this.formula = parse(formula);
        var parts =angular.copy(this.parts);
        var gormular=angular.copy(this.formula);
        if (merge) {
            parts =angular.copy( merge.parts && merge.parts.map(responsePart) || []);
            formula=angular.copy(parse(merge.formula));
        }

        this.declaredVars = uniqueStrings(
            parts.reduce(function(prev, part) {
                return prev.concat(part.declaredVars);
            }, []));
        this.freeVars = uniqueStrings(
            parts.reduce(function(prev, part) {
                return prev.concat(part.freeVars);
            }, refs(formula)));
        this.unboundVars = uniqueStrings(
            this.freeVars.filter(function(fv) {
                return this.declaredVars.indexOf(fv) < 0;
            }, this));
    };*/

    function ResponseType(name, partsDef, formula, merge) {
        this.name = name;
        
        if (merge) {
            //this.parts = merge.parts && merge.parts.map(responsePart) || [];
            this.parts = partsDef && partsDef.map(responsePart) || [];
           this.formula=parse(merge.formula);
        }
        else {
            this.parts = partsDef && partsDef.map(responsePart) || [];
            this.formula = parse(formula);
        }
        if (this.formula && this.formula.functions) {
            this.formula.functions.median = function(...arr) {
                const mid = Math.floor(arr.length / 2),
                    nums = [...arr].sort((a, b) => a - b);
                return arr.length % 2 !== 0 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2;
            }
        }
        this.declaredVars = uniqueStrings(
            this.parts.reduce(function(prev, part) {
                return prev.concat(part.declaredVars);
            }, []));
        this.freeVars = uniqueStrings(
            this.parts.reduce(function(prev, part) {
                return prev.concat(part.freeVars);
            }, refs(this.formula)));
        this.unboundVars = uniqueStrings(
            this.freeVars.filter(function(fv) {
                return this.declaredVars.indexOf(fv) < 0;
            }, this));
    };


    
    ResponseType.prototype.zip = function(responseParts) {
        var partPairs = [];
        for (var i = 0; i < this.parts.length; i++) {
            partPairs.push({
                index: i,
                schema: this.parts[i],
                data: responseParts[i],
            });
        }
        return partPairs;
    };
    ResponseType.prototype.score = function(responseParts, scope) {
        if (this.formula)
            return this.formula.evaluate(scope);

        var score = 0.0;
        this.zip(responseParts).forEach(function(part) {
            score += part.schema.score(part.data);
        });
        return score;
    };
    ResponseType.prototype.variables = function(responseParts, ignoreErrors) {
        var scope = {};
        this.zip(responseParts).forEach(function(part) {
            var vars;
            try {
                vars = part.schema.variables(part.data);
            } catch (e) {
                if (!ignoreErrors)
                    throw e;
            }
            angular.merge(scope, vars);
        });
        return scope;
    };
    ResponseType.prototype.humanize_variable = function(field_name) {
        if (field_name == '_raw')
            return 'Raw score';
        if (field_name == '_score')
            return 'Weighted score';
        if (field_name == '_weight')
            return 'Measure weight';
        for (var i = 0; i < this.parts.length; i++) {
            var part = this.parts[i];
            if (field_name != part.name)
                continue;
            if (part.name)
                return 'Part ' + part.name;
            return 'Part ' + i;
        }
        return 'Unknown field ' + field_name;
    };
    ResponseType.prototype.validate = function(responseParts, scope) {
        this.zip(responseParts).forEach(function(part, index) {
            try {
                part.schema.validate(part.data, scope);
            } catch (e) {
                var name;
                if (part.schema.name) {
                    name = '' + part.schema.name;
                } else {
                    if (this.parts.length > 1)
                        name = "Part " + (index + 1);
                    else
                        name = "Response";
                }
                throw name + ": " + e;
            }
        }, this);
    };

    //check responses depend on choose by detang on 17/09/2019
    ResponseType.prototype.validateResponses = function(responseParts, scope) {
        this.zip(responseParts).forEach(function(part, index) {
            try {
                part.schema.validate(part.data, scope);
            } catch (e) {
              //empty response if choose make it invalid 
              delete part.data.index;
              delete part.data.note;
            }
        }, this);
    };


    ResponseType.prototype.calculate_score = function(responseParts) {
        if (!responseParts || responseParts.length != this.parts.length)
            throw "Response is incomplete.";

        var scope = this.variables(responseParts);
        this.validate(responseParts, scope);
        return this.score(responseParts, scope);
    };


    function responsePart(pDef) {
        if (pDef.type == 'multiple_choice')
            return new MultipleChoice(pDef);
        else
            return new Numerical(pDef);
    };


    function ResponsePart(pDef) {
        this.id = pDef.id;
        this.name = pDef.name;
        this.type = pDef.type;
        this.description = pDef.description;
        if (pDef.submeasure) {
            this.submeasure=pDef.submeasure;
        }
    };


    function MultipleChoice(pDef) {
        ResponsePart.call(this, pDef);
        this.options = pDef.options.map(function(oDef) {
            return new ResponseOption(oDef);
        });
        this.declaredVars = this.id ? [this.id, this.id + '__i'] : [];
        this.freeVars = uniqueStrings(
            this.options.reduce(function(prev, option) {
                return prev.concat(option.freeVars);
            }, []));
    };
    MultipleChoice.prototype = Object.create(ResponsePart.prototype);
    MultipleChoice.prototype.score = function(part) {
        var option = this.options[part.index];
        return option.score;
    };
    MultipleChoice.prototype.variables = function(part) {
        var variables = {};
        if (this.id) {
            variables[this.id] = this.options[part.index].score;
            variables[this.id + '__i'] = part.index;
        }
        return variables;
    };
    MultipleChoice.prototype.validate = function(part, scope) {
        if (part.index == null)
            throw "Choose an option";
        var option = this.options[part.index];
        if (!option.available(scope))
            throw "Can't select \"" + option.name + "\"";
    };


    function ResponseOption(oDef) {
        this.score = oDef.score;
        this.name = oDef.name;
        this.description = oDef.description;
        this.predicate = parse(oDef['if']);
        this.freeVars = refs(this.predicate);
    };
    ResponseOption.prototype.available = function(scope) {
        if (!this.predicate)
            return true;
        try {
            return Boolean(this.predicate.evaluate(scope));
        } catch (e) {
            return false;
        }
    };


    function Numerical(pDef) {
        ResponsePart.call(this, pDef);
        this.lowerExp = parse(pDef.lower);
        this.upperExp = parse(pDef.upper);
        this.declaredVars = this.id ? [this.id] : [];
        this.freeVars = uniqueStrings(
            refs(this.lowerExp).concat(refs(this.upperExp)));
    };
    Numerical.prototype = Object.create(ResponsePart.prototype);
    Numerical.prototype.score = function(part) {
        return part.value;
    };
    Numerical.prototype.variables = function(part) {
        var variables = {};
        if (this.id)
            variables[this.id] = part.value;
        return variables;
    };
    Numerical.prototype.lower = function(scope) {
        if (!this.lowerExp)
            return Number.NEGATIVE_INFINITY;
        return this.lowerExp.evaluate(scope);
    };
    Numerical.prototype.upper = function(scope) {
        if (!this.upperExp)
            return Number.POSITIVE_INFINITY;
        return this.upperExp.evaluate(scope);
    };
    Numerical.prototype.validate = function(part, scope) {
        if (part.value == null)
            throw "Please specify a value";
        var score = this.score(part);
        try {
            if (this.lower(scope) > score)
                throw "Must be at least " + this.lower(scope);
            if (this.upper(scope) < score)
                throw "Must be at most " + this.upper(scope);
        } catch (e) {
            throw e;
        }
    };

    return {
        ResponseType: ResponseType,
        MultipleChoice: MultipleChoice,
        Numerical: Numerical,
    };
})


.factory('ResponseType', ['$resource', 'paged', function($resource, paged) {
    var ResponseType = $resource('/response_type/:id.json', {
        id: '@id', programId: '@programId'
    }, {
        get: {
            method: 'GET', cache: false,
            interceptor: {response: function(response) {
                response.resource.title = response.resource.name;
                return response.resource;
            }}
        },
        create: { method: 'POST' },
        save: { method: 'PUT' },
        query: {
            method: 'GET', isArray: true, cache: false,
            interceptor: {response: paged}
        },
        history: { method: 'GET', url: '/response_type/:id/program.json',
            isArray: true, cache: false }
    });
    ResponseType.prototype.$createOrSave = function(parameters, success, error) {
        if (!this.id)
            return this.$create(parameters, success, error);
        else
            return this.$save(parameters, success, error);
    };
    return ResponseType;
}])


.controller('ResponseTypeCtrl',
        function($scope, Authz, Measure, layout, routeData,
            ResponseType) {

    $scope.layout = layout;
    $scope.checkRole = Authz({});
    $scope.responseType = routeData.responseType;
    $scope.ResponseType = ResponseType;
})


.directive('responseTypeEditor', function() {
    return {
        restrict: 'A',
        scope: {
            rt: '=responseTypeEditor',
            weight: '=',
            isBound: '=',
        },
        templateUrl: 'response_type_editor.html',
        controller: function($scope, Numbers, responseTypes, $timeout, Enqueue) {
            //$scope.$on("remove-responseType",function(){
            //    $scope.$broadcast("remove-responseType");
            //})
            $scope.$watch('rt', function(rt) {
                $scope.rtEdit = {
                    rt: rt,
                    responseType: null,
                    externs: {},
                    response: {
                        responseParts: [],
                        comment: ''
                    },
                    activeTab: 'details',
                };
            });
            $scope.addPart = function(rt, $event) {
                var ids = {};
                for (var i = 0; i < rt.parts.length; i++) {
                    ids[rt.parts[i].id] = true;
                }
                var id;
                for (var i = 0; i <= rt.parts.length; i++) {
                    id = Numbers.idOf(i);
                    if (!ids[id])
                        break;
                }
                var part = {
                    id: id,
                    name: 'Response part ' + id.toUpperCase(),
                };
                $scope.setType(part, 'multiple_choice');
                rt.parts.push(part);
                $scope.updateFormula(rt);

                $timeout(function() {
                    $scope.rtEdit.activeTab = rt.parts.indexOf(part);
                });
            };
            $scope.setType = function(part, type) {
                part.type = type;
                if (type == 'multiple_choice' && !part.options) {
                    part.options = [
                        {score: 0, name: 'No', 'if': null},
                        {score: 1, name: 'Yes', 'if': null}
                        //{score: 0.5, name: 'Yes', 'if': null}
                    ];
                }
            };
            $scope.addOption = function(part) {
                part.options.push({
                    score: 0,
                    //score: part.options.length / (part.options.length + 1),
                    name: 'Option ' + (part.options.length + 1)
                })
            };
            $scope.updateFormula = function(rt) {
                if (rt.parts.length <= 1) {
                    rt.formula = null;
                    return;
                }
                var formula = "";
                for (var i = 0; i < rt.parts.length; i++) {
                    var part = rt.parts[i];
                    if (i > 0)
                        formula += " + ";
                    if (part.id)
                        formula += part.id;
                    else
                        formula += "?";
                }
                rt.formula = formula;
            };
            $scope.remove = function(rt, list, item) {
                var i = list.indexOf(item);
                if (i < 0)
                    return;
                list.splice(i, 1);
                if (item.options)
                    $scope.updateFormula(rt);
            };

            var rtDefChanged = Enqueue(function() {
                var rtDef = $scope.rtEdit.rt;
                if (!rtDef) {
                    $scope.rtEdit.responseType = null;
                    return;
                }
                //var merge=null;
                //if ($scope.$root && $scope.$root.merge) {
                //    merge=$scope.$root.merge;
                //}
                $scope.rtEdit.responseType = new responseTypes.ResponseType(
                    rtDef.name, rtDef.parts, rtDef.formula);
            }, 0, $scope);
            $scope.$watch('rtEdit.rt', rtDefChanged);
            $scope.$watch('rtEdit.rt', rtDefChanged, true);

            $scope.partTypes = [
                {name: 'multiple_choice', desc: 'Multiple choice'},
                {name: 'numerical', desc: 'Numerical'},
            ];
            $scope.partTypeMap = $scope.partTypes.reduce(function(ts, t){
                ts[t.name] = t.desc;
                return ts;
            }, {});

            $scope.$watchGroup(['rt.nMeasures', 'isBound'], function(vars) {
                var nMeasures = vars[0];
                if ($scope.isBound)
                    $scope.nMeasures = nMeasures - 1;
                else
                    $scope.nMeasures = nMeasures;
            });
        },
        link: function(scope, elem, attrs) {
        },
    };
})


.directive('responseTypesEditor', function() {
    return {
        restrict: 'A',
        scope: {
            rt: '=responseTypesEditor',
            indexSub:'=indexSub',
            formula:'=',
            weight: '=',
            isBound: '=',
        },
        templateUrl: 'response_types_editor.html',
        controller:  function($scope, Numbers, responseTypes, $timeout, Enqueue, $rootScope) {
            if ($scope.indexSub || $scope.indexSub==0) {
                if ($rootScope.rts && $rootScope.rts[$scope.indexSub]) {
                    $rootScope.rts[$scope.indexSub]=$scope;
                    //$rootScope.questions.push($scope);
                }
                else {
                    //$rootScope.questions=[$scope];
                }
            }
            $scope.$watch('rt', function(rt) {
                $scope.rtEdit = {
                    rt: rt,
                    responseType: null,
                    externs: {},
                    response: {
                        responseParts: [],
                        comment: ''
                    },
                    activeTab: 'details',
                };
            });
            $scope.addPart = function(rt, $event) {
                var ids = {};
                for (var i = 0; i < rt.parts.length; i++) {
                    ids[rt.parts[i].id] = true;
                }
                var id;
                for (var i = 0; i <= rt.parts.length; i++) {
                    id = Numbers.idOf(i);
                    if (!ids[id])
                        break;
                }
                var part = {
                    id: id,
                    name: 'Response part ' + id.toUpperCase(),
                };
                $scope.setType(part, 'multiple_choice');
                rt.parts.push(part);
                $scope.updateFormula(rt);

                $timeout(function() {
                    $scope.rtEdit.activeTab = rt.parts.indexOf(part);
                });
            };
            $scope.setType = function(part, type) {
                part.type = type;
                if (type == 'multiple_choice' && !part.options) {
                    part.options = [
                        {score: 0, name: 'No', 'if': null},
                        {score: 1, name: 'Yes', 'if': null}
                        //{score: 0.5, name: 'Yes', 'if': null}
                    ];
                }
            };
            $scope.addOption = function(part) {
                part.options.push({
                    score: 0,
                    //score: part.options.length / (part.options.length + 1),
                    name: 'Option ' + (part.options.length + 1)
                })
            };
            $scope.updateFormula = function(rt) {
                if (rt.parts.length <= 1) {
                    rt.formula = null;
                    return;
                }
                var formula = "";
                for (var i = 0; i < rt.parts.length; i++) {
                    var part = rt.parts[i];
                    if (i > 0)
                        formula += " + ";
                    if (part.id)
                        formula += part.id;
                    else
                        formula += "?";
                }
                rt.formula = formula;
            };
            $scope.remove = function(rt, list, item) {
                var i = list.indexOf(item);
                if (i < 0)
                    return;
                list.splice(i, 1);
                if (item.options)
                    $scope.updateFormula(rt);
            };

            var rtDefChanged = Enqueue(function() {
                //keep submeasure index
                if ($scope.indexSub)
                   $rootScope.indexSub=$scope.indexSub; 
                var rtDef = $scope.rtEdit.rt;
                if (!rtDef) {
                    $scope.rtEdit.responseType = null;
                    return;
                }
                var mergeParts=[];
                angular.forEach($scope.$root.rts,function(rt,i){
                    if (!angular.equals({}, rt)) {
                       mergeParts=[...mergeParts, ...rt.rtEdit.rt.parts]
                    }
                });
                var merge={parts:mergeParts, formula:$scope.formula};
                $scope.rtEdit.responseType = new responseTypes.ResponseType(
                    rtDef.name, rtDef.parts, rtDef.formula, merge);
                $scope.$emit('response-type-changed');
                angular.forEach($scope.$root.rts,function(s,i){
                    if (!angular.equals({},s)) { 
                        if (s.rtEdit.activeTab == 'preview')               
                            s.rtEdit.activeTab = 'details';
                    }
                })
            }, 0, $scope);
            $scope.$watch('rtEdit.rt', rtDefChanged);
            $scope.$watch('rtEdit.rt', rtDefChanged, true);
            


            /*$scope.$watch('formula',  function(vars) {
                if ($scope.indexSub || $rootScope.indexSub) {
                    return rtDefChanged;
                }
            }, true);*/ 

            $scope.$watch('formula', rtDefChanged);

            $scope.$watch('formula', rtDefChanged, true);





            $scope.partTypes = [
                {name: 'multiple_choice', desc: 'Multiple choice'},
                {name: 'numerical', desc: 'Numerical'},
            ];
            $scope.partTypeMap = $scope.partTypes.reduce(function(ts, t){
                ts[t.name] = t.desc;
                return ts;
            }, {});

            $scope.$watchGroup(['rt.nMeasures', 'isBound'], function(vars) {
                var nMeasures = vars[0];
                if ($scope.isBound)
                    $scope.nMeasures = nMeasures - 1;
                else
                    $scope.nMeasures = nMeasures;
            });

            //$scope.loadFooItems= function(rt, list, item) {
            //    $scope.rtEdit.activeTab = "preview";
               // var i = list.indexOf(item);
                //if (i < 0)
                 //   return;
                //list.splice(i, 1);
                //if (item.options)
                //    $scope.updateFormula(rt);
            //};
            //this.setActiveTab = function() { 
            //    $scope.rtEdit.activeTab = "preview";
            //};
        },
        link: function(scope, elem, attrs) {
            //ngCtrl.setActiveTab();
            //$scope.setActiveTab();
            //var notSetActiveTab=true;
            /*if (scope.$parent.$parent.$parent.$parent.previewList) {
                scope.$parent.$parent.$parent.$parent.previewList.push(scope);
            }
            else {
                scope.$parent.$parent.$parent.$parent.previewList=[scope];
            }*/
             
            scope.setActiveTab = function(tab) { 
                if (scope.$root.indexSub && scope.$root.rts) {
                    var lastRt=0;
                    var unboundVars=[];
                    var indexSub=scope.$root.indexSub;
                    angular.forEach(scope.$root.rts,function(s,i){
                        if (!angular.equals({},s)) {                
                            s.rtEdit.activeTab = tab;
                            if (indexSub != i && s.rtEdit.responseType.unboundVars.length>0) {
                            s.rtEdit.responseType.unboundVars=[];
                        }
                            lastRt=i;
                        }
                    })
                //only display in the last no empty submeasure
                
                    if (indexSub != lastRt && scope.$root.rts[indexSub].rtEdit.responseType.unboundVars.length>0) {
                       unboundVars=angular.copy(scope.$root.rts[indexSub].rtEdit.responseType.unboundVars);
                       scope.$root.rts[indexSub].rtEdit.responseType.unboundVars=[];
                       scope.$root.rts[lastRt].rtEdit.responseType.unboundVars=angular.copy(unboundVars);
                    }

                }
                /*scope.$root.rts[lastRt].rtEdit.responseType.unboundVars=unboundVars;
                //if (setActiveTab) {
                    angular.forEach(scope.$parent.$parent.$parent.$parent.previewList,function(s,i){
                        //if (tab!='')
                           s.rtEdit.activeTab = tab;
                        //else
                        //   s.rtEdit.activeTab = 'details'; 
                    })*/
            }
                
           // };           
            //console.log(elem.controller('responseTypesEditor'));
        },
    };
})

;
